tools/base/aio.py

```diff
@@ -83,3 +105,405 @@ async def run_system_command():
+
+    def forget_task(self, task: asyncio.Task) -> None:
+        """Task? what task?"""
+        if self.closed:
+            # If we are closing, don't remove, as this has been triggered
+            # by cancellation.
+            return
+        self.running_tasks.remove(task)
+
+    async def iter_coros(self) -> AsyncIterator[Union[AsyncParallelIteratorError, Awaitable]]:
+        """Iterate provided coros either synchronously or asynchronously,
+        yielding the awaitables asynchoronously.
+        """
+        try:
+            if self.is_async:
+                async for coro in self._coros:  # type:ignore
+                    yield coro
+            else:
+                for coro in self._coros:  # type:ignore
+                    yield coro
+        except BaseException as e:
+            # Catch all errors iterating (other errors are caught elsewhere)
+            # If iterating raises, wrap the error and send it to `submit` and
+            # and `output` to close the queues.
+            yield AsyncParallelIteratorError(e)
+
+    async def on_task_complete(self, result: Any, decrement: Optional[bool] = True) -> None:
+        """Output the result, release the sem lock, decrement the running
+        count, and notify output queue if complete.
+        """
+        if self.closed:
+            # Results can come back after the queue has closed as they are
+            # cancelled.
+            # In that case, nothing further to do.
+            return
+
+        # Give result to output
+        await self.out.put(result)
+
+        if not self.nolimit:
+            # Release the sem.lock
+            self.sem.release()
+        if decrement:
+            # Decrement the running_queue if it was incremented
+            self.running_queue.get_nowait()
+        # Exit if nothing left to do
+        await self.exit_on_completion()
+
+    async def output(self) -> AsyncIterator:
+        """Asynchronously yield results as they become available"""
+        while True:
+            # Wait for some output
+            result = await self.out.get()
+            if result is _marker:
+                # All done!
+                await self.close()
+                break
+            elif self.should_error(result):
+                # Raise an error and bail!
+                await self.cancel()
+                raise result
+            yield result
+
+    async def ready(self) -> bool:
+        """Wait for the sem.lock and indicate availability in the submission
+        queue
+        """
+        if self.closed:
+            return False
+        if not self.nolimit:
+            await self.sem.acquire()
+            # We check before and after acquiring the sem.lock to see whether
+            # we are `closed` as these events can be separated in
+            # time/procedure.
+            if self.closed:
+                return False
+        return True
+
+    def remember_task(self, task: asyncio.Task) -> None:
```

Comment:
It's a bit hard to distinguish internal vs. external code here. Can you "privatize" methods that are not generally supposed to be part of the public interface?